File: src/api/schemas/diagnosis.py
"""
Pydantic Request & Response Schemas for Phase 9 FastAPI Service.
"""

from typing import Any

from pydantic import BaseModel, Field


# --- Health & Readiness Schemas ---
class HealthResponse(BaseModel):
    status: str = Field(json_schema_extra={"example": "healthy"})
    service: str = Field(json_schema_extra={"example": "ai-field-engineer-api"})
    version: str = Field(json_schema_extra={"example": "1.0.0"})
    environment: str = Field("production", json_schema_extra={"example": "production"})
    git_sha: str = Field("unknown", json_schema_extra={"example": "5963611"})
    timestamp: str


class ReadinessResponse(BaseModel):
    ready: bool
    status: str
    components: dict[str, bool] = Field(
        json_schema_extra={
            "example": {
                "vision_service": True,
                "audio_service": True,
                "sensor_service": True,
                "multimodal_fusion": True,
                "rag_retrieval": True,
                "diagnostic_agent": True,
                "explainability_service": True,
            }
        }
    )
    timestamp: str


# --- Sensor & Telemetry Input Schemas ---
class SensorTelemetryInput(BaseModel):
    temperature: float | None = Field(
        None, description="Bearing or casing temperature", json_schema_extra={"example": 84.0}
    )
    temperature_unit: str = Field("degC", description="Temperature unit (degC or degF)")
    vibration: float | None = Field(None, description="RMS vibration velocity", json_schema_extra={"example": 6.8})
    vibration_unit: str = Field("mm/s", description="Vibration unit (mm/s or in/s)")
    rpm: float | None = Field(
        None, description="Operating shaft rotational speed", json_schema_extra={"example": 1480.0}
    )
    current: float | None = Field(None, description="Motor operating current", json_schema_extra={"example": 8.4})
    current_unit: str = Field("A", description="Current unit")
    pressure: float | None = Field(
        None, description="Suction or discharge pressure", json_schema_extra={"example": 1.2}
    )
    pressure_unit: str = Field("bar", description="Pressure unit")
    custom_parameters: dict[str, float] = Field(
        default_factory=dict, description="Additional arbitrary sensor readings"
    )


# --- Equipment Metadata ---
class EquipmentMetadataInput(BaseModel):
    equipment_type: str = Field(
        default="motor",
        description="Equipment classification (e.g. motor, pump, gearbox)",
        json_schema_extra={"example": "motor"},
    )
    manufacturer: str | None = Field(
        default=None, description="Equipment manufacturer", json_schema_extra={"example": "Siemens"}
    )
    model: str | None = Field(
        default=None, description="Equipment model identifier", json_schema_extra={"example": "M-4500"}
    )
    serial_number: str | None = Field(
        default=None, description="Asset serial number", json_schema_extra={"example": "SN-94821"}
    )
    operating_mode: str | None = Field(default="continuous", description="Duty cycle mode")


# --- Evidence & Action Output Schemas ---
class EvidenceItemResponse(BaseModel):
    evidence_id: str
    category: str
    source: str
    description: str
    quality: str
    raw_value: float | None = None
    unit: str | None = None
    document_name: str | None = None
    page_number: int | None = None
    section: str | None = None


class RecommendedActionResponse(BaseModel):
    action_id: str
    priority: int
    requirement: str
    action_text: str
    rationale: str
    source_reference: str | None = None
    is_safety_critical: bool
    justifying_evidence_ids: list[str] = Field(default_factory=list)


class ClaimAuditMappingResponse(BaseModel):
    claim_id: str
    claim_statement: str
    status: str
    supporting_evidence_ids: list[str]
    contradicting_evidence_ids: list[str]
    rationale: str | None = None


class ConfidenceDecompositionResponse(BaseModel):
    overall_confidence: float
    multimodal_agreement: str
    sensor_evidence_strength: str
    acoustic_evidence_strength: str
    visual_evidence_strength: str
    technical_knowledge_match: str
    contradiction_penalty: float
    rationale_summary: str


class PrimaryDiagnosisResponse(BaseModel):
    primary_diagnosis: str
    diagnostic_confidence: float
    severity: str
    confidence_decomposition: ConfidenceDecompositionResponse


# --- Main Diagnostic Response Schema ---
class DiagnosisResponse(BaseModel):
    case_id: str
    request_id: str
    timestamp: str
    status: str
    equipment: dict[str, Any]
    problem_summary: str
    available_modalities: list[str]
    diagnosis: PrimaryDiagnosisResponse
    evidence_inventory: list[EvidenceItemResponse]
    claim_mappings: list[ClaimAuditMappingResponse]
    alternative_hypotheses: list[dict[str, Any]]
    recommended_actions: list[RecommendedActionResponse]
    uncertainty_profile: dict[str, Any]
    unsupported_claims: list[str]
    audit_summary: dict[str, Any]
    markdown_report: str


# --- Error Response Schema ---
class ErrorResponse(BaseModel):
    error_code: str
    message: str
    details: dict[str, Any] | None = None
    request_id: str | None = None
    timestamp: str

File: frontend/src/lib/types/diagnosis.ts
/**
 * TypeScript Type Definitions Synchronized Exactly with Phase 9 FastAPI Pydantic Models.
 */

export interface HealthResponse {
  status: string;
  service: string;
  version: string;
  environment: string;
  git_sha: string;
  timestamp: string;
}

export interface ReadinessResponse {
  ready: boolean;
  status: string;
  components: Record<string, boolean>;
  timestamp: string;
}

export interface SensorTelemetryInput {
  temperature?: number;
  temperature_unit?: string;
  vibration?: number;
  vibration_unit?: string;
  rpm?: number;
  current?: number;
  current_unit?: string;
  pressure?: number;
  pressure_unit?: string;
  custom_parameters?: Record<string, number>;
}

export interface EquipmentMetadataInput {
  equipment_type: string;
  manufacturer?: string;
  model?: string;
  serial_number?: string;
  operating_mode?: string;
}

export interface EvidenceItemResponse {
  evidence_id: string;
  category: string;
  source_modality: string;
  description: string;
  confidence: float_number;
  raw_measurement?: string;
  iso_standard_reference?: string;
  document_reference?: string;
  page_number?: number;
  is_verified_citation: boolean;
}

type float_number = number;

export interface ConfidenceDecompositionResponse {
  overall_confidence: number;
  multimodal_agreement: string;
  sensor_evidence_strength: string;
  acoustic_evidence_strength: string;
  visual_evidence_strength: string;
  technical_knowledge_match: string;
  contradiction_penalty: number;
  rationale_summary: string;
}

export interface PrimaryDiagnosisResponse {
  primary_diagnosis: string;
  diagnostic_confidence: number;
  severity: string;
  confidence_decomposition: ConfidenceDecompositionResponse;
}

export interface RecommendedActionResponse {
  action_id: string;
  action_type: string;
  description: string;
  target_component: string;
  justifying_evidence_ids: string[];
  citation?: string;
}

export interface DiagnosisResponse {
  case_id: string;
  request_id: string;
  timestamp: string;
  status: string;
  equipment: Record<string, any>;
  problem_summary: string;
  available_modalities: string[];
  diagnosis: PrimaryDiagnosisResponse;
  evidence_inventory: EvidenceItemResponse[];
  claim_mappings: any[];
  alternative_hypotheses: any[];
  recommended_actions: RecommendedActionResponse[];
  uncertainty_profile: Record<string, any>;
  unsupported_claims: string[];
  audit_summary: Record<string, any>;
  markdown_report: string;
}

export interface RAGQueryMatch {
  chunk_id: string;
  document_name: string;
  page_number?: number;
  section?: string;
  similarity_score: number;
  content_preview: string;
}

export interface RAGQueryResponse {
  query: string;
  results_count: number;
  matches: RAGQueryMatch[];
  timestamp: string;
}
